active/0000-seme-regions.md

```diff
@@ -0,0 +1,305 @@
+- Start Date: 2014-10-12
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+
+The regions used for lifetime inference and borrow checking should be general
+single-entry / multiple-exit regions, instead of being limited to the current
+situation of lexical scopes, or even more general single-entry / single-exit
+regions.
+
+# Motivation
+
+There are three classes of program fragments that are currently rejected by the
+borrow checker that really seem like they should be accepted.
+
+1. The borrow checker ignores liveness of borrows, so that a borrow is valid for
+the entirety of a lexical scope, even after its last use. This even occurs in
+straight-line code For example, a program like this is rejected:
+
+    ```rust
+    let mut i = 0i;
+    let p = &mut i;
+    *p = 1;
+    let q = &mut i;
+    *q = 2;
+    ```
+
+ The borrow checker should be able to realize that each of these borrows is no
+longer valid at the point of the next borrow. The workaround for this is to give
+each borrow its own lexical scope, e.g.
+
+    ```rust
+    let mut i = 0i;
+    {
+        let p = &mut i;
+        *p = 1;
+    }
+    {
+        let q = &mut i;
+        *q = 2;
+    }
+    ```
+
+ This is unnecesarily verbose and difficult to explain to new programmers. It
+also doesn't handle all cases involving mutable borrows of multiple locations.
+
+2. The borrow checker can't refine borrow information based on the branches of
+a pattern match. The canonical example is a 'find-or-insert' pattern:
+
+    ```rust
+    match map.find(&key) {
+        Some(...) => { ... }
+        None => {
+            map.insert(key, new_value);
+        }
+    }
+    ```
+
+ In an `Option<&T>` the `&T` only appears in the the `Some` constructor, so the
+borrow shouldn't need to be considered in the `None` branch. The workaround for
+this is usually to do a clone/copy and two separate lookups'. This is more
+verbose, doesn't work in all cases, and is potentially a large performance
+penalty.
+
+3. Borrows in rvalues of the same overall expression often overlap when it
+doesn't seem that they have to. For example, if `a` has methods `foo`, `bar`
+and `baz` that all take `&mut self` parameters, but the latter two have no
+lifetimes in output position, then an expression such as
+
+    ```rust
+    a.foo(a.bar(), a.baz())
+    ```
+
+ seems like it should be valid. The workaround is to instead write
+
+    ```rust
+    let bar = a.bar();
+    let baz = a.baz();
+    a.foo(bar, baz);
+    ```
+
+ All this is doing is making the evaluation of temporaries explicit, so it
```

Comment:
Is there a reason you want to go with this specific evaluation order? A Google search did not produce any documentation on this.

Your workaround would work for simple cases. It will break apart unpredictably at more complex cases. Such unpredictably would really frustrate the users.
What about debug mode with optimizations disabled? Would it still reorder stuff?
 
What to do if you need a mutable borrow to evaluate parameters:

    obj.paint(obj.x, obj.y, obj.getCachedOrDownloadImage())

Or a mutable borrow to get to the function to call:

    obj.getSomethingMutablyBorrowsObj().someFunction(obj.name)  

To generalize self is just another parameter. If more than one parameter needs to borrow ob after a mutable borrow is done, there is no way out.

We can either define an order that would work for most cases or leave it undefined and let compiler to figure out if there is a evaluation order that complies with borrow restrictions.

It's much more likely that self will need a mutable borrow of obj than the other parameters. 
Thus I'm for evaluating self last to allow more code to work.